File: codes/python/leetcodes/src/main/python/com/hit/basmath/learn/array_and_string/_54.py
"""
54. Spiral Matrix

Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.

Example 1:

Input:
[
 [ 1, 2, 3 ],
 [ 4, 5, 6 ],
 [ 7, 8, 9 ]
]
Output: [1,2,3,6,9,8,7,4,5]

Example 2:

Input:
[
  [1, 2, 3, 4],
  [5, 6, 7, 8],
  [9,10,11,12]
]
Output: [1,2,3,4,8,12,11,10,9,5,6,7]
"""


class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        targetList = []

        if len(matrix) == 0:
            return targetList

        rowBegin = 0
        rowEnd = len(matrix) - 1
        colBegin = 0
        colEnd = len(matrix[0]) - 1

        while rowBegin <= rowEnd and colBegin <= colEnd:
            i = colBegin
            while i <= colEnd:
                targetList.append(matrix[rowBegin][i])
                i += 1

            rowBegin += 1

            j = rowBegin
            while j <= rowEnd:
                targetList.append(matrix[j][colEnd])
                j += 1

            colEnd -= 1

            if rowBegin <= rowEnd:
                k = colEnd
                while k >= colBegin:
                    targetList.append(matrix[rowEnd][k])
                    k -= 1

            rowEnd -= 1

            if colBegin <= colEnd:
                m = rowEnd
                while m >= rowBegin:
                    targetList.append(matrix[m][colBegin])
                    m -= 1

            colBegin += 1

        return targetList

File: codes/java/leetcodes/src/main/java/com/hit/basmath/learn/array_and_string/_54.java
package com.hit.basmath.learn.array_and_string;

import java.util.ArrayList;
import java.util.List;

/**
 * 54. Spiral Matrix
 * <p>
 * Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.
 * <p>
 * For example, given the following matrix:
 * <p>
 * [
 * [ 1, 2, 3 ],
 * [ 4, 5, 6 ],
 * [ 7, 8, 9 ]
 * ]
 * <p>
 * You should return [1,2,3,6,9,8,7,4,5].
 */
public class _54 {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> targetList = new ArrayList<Integer>();

        // If matrix.length == 0, return targetList immediately
        if (matrix.length == 0) return targetList;

        // Initial variables
        int rowBegin = 0, rowEnd = matrix.length - 1;
        int colBegin = 0, colEnd = matrix[0].length - 1;

        // Iterate all element of matrix，condition is rowBegin <= rowEnd && colBegin <= colEnd
        while (rowBegin <= rowEnd && colBegin <= colEnd) {
            // Traverse Right
            for (int i = colBegin; i <= colEnd; i++) {
                targetList.add(matrix[rowBegin][i]);
            }

            /**
             * When first circle finished, it's mean first row has handled over,
             * so rowBegin++, and the next cycle is similar.
             */
            rowBegin++;

            // Traverse Down
            for (int j = rowBegin; j <= rowEnd; j++) {
                targetList.add(matrix[j][colEnd]);
            }

            /**
             * When this circle finished, it's mean last column has handled over,
             * so colEnd--, and the next cycle is similar.
             */
            colEnd--;

            if (rowBegin <= rowEnd) {
                // Traverse Left
                for (int k = colEnd; k >= colBegin; k--) {
                    targetList.add(matrix[rowEnd][k]);
                }
            }

            /**
             * When this circle finished, it's mean last row has handled over,
             * so rowEnd--, and the next cycle is similar.
             */
            rowEnd--;

            if (colBegin <= colEnd) {
                // Traver Up
                for (int m = rowEnd; m >= rowBegin; m--) {
                    targetList.add(matrix[m][colBegin]);
                }
            }

            /**
             * When this circle finished, it's mean first row has handled over,
             * so colBegin++, and the next cycle is similar.
             */
            colBegin++;
        }
        return targetList;
    }
}
